deimos/openssl/stack.d

```diff
@@ -76,26 +76,92 @@ alias stack_st _STACK;  /* Use STACK_OF(...) instead */
 int M_sk_num()(_STACK* sk) { return (sk ? sk.num : -1); }
 char* M_sk_value()(_STACK* sk, size_t n) { return (sk ? sk.data[n] : null); }
 
-int sk_num(const(_STACK)*);
-void* sk_value(const(_STACK)*, int);
-
-void* sk_set(_STACK*, int, void*);
-
-_STACK* sk_new(ExternC!(int function(const(void)*, const(void)*)) cmp);
-_STACK* sk_new_null();
-void sk_free(_STACK*);
-void sk_pop_free(_STACK* st, ExternC!(void function(void*)) func);
-int sk_insert(_STACK* sk, void* data, int where);
-void* sk_delete(_STACK* st, int loc);
-void* sk_delete_ptr(_STACK* st, void* p);
-int sk_find(_STACK* st, void* data);
-int sk_find_ex(_STACK* st, void* data);
-int sk_push(_STACK* st, void* data);
-int sk_unshift(_STACK* st, void* data);
-void* sk_shift(_STACK* st);
-void* sk_pop(_STACK* st);
-void sk_zero(_STACK* st);
-int function(const(void)*, const(void)*) sk_set_cmp_func(_STACK* sk, ExternC!(int function(const(void)*, const(void)*)) c);
-_STACK* sk_dup(_STACK* st);
-void sk_sort(_STACK* st);
-int sk_is_sorted(const(_STACK)* st);
+/*******************************************************************************
+
+    Starting from OpenSSL v1.1.0, the `sk_*` functions are prefixed with
+    `OPENSSL_`, so e.g. `sk_num` becomes `OPENSSL_sk_num`.
+
+    To cope with that and provide downstream with an easier interface to deal
+    with, we always provide the `OPENSSL_` methods, as they are just `extern(C)`
+    anyway, and we either provide `sk_*` as `alias`es or `extern (C)`.
+
+    The need for aliases comes from users of those binding, e.g. `safestack`.
+
+*******************************************************************************/
+
+int OPENSSL_sk_num(const(_STACK)*);
+void* OPENSSL_sk_value(const(_STACK)*, int);
+
+void* OPENSSL_sk_set(_STACK*, int, void*);
+
+_STACK* OPENSSL_sk_new(ExternC!(int function(const(void)*, const(void)*)) cmp);
+_STACK* OPENSSL_sk_new_null();
+void OPENSSL_sk_free(_STACK*);
+void OPENSSL_sk_pop_free(_STACK* st, ExternC!(void function(void*)) func);
+int OPENSSL_sk_insert(_STACK* sk, void* data, int where);
+void* OPENSSL_sk_delete(_STACK* st, int loc);
+void* OPENSSL_sk_delete_ptr(_STACK* st, void* p);
+int OPENSSL_sk_find(_STACK* st, void* data);
+int OPENSSL_sk_find_ex(_STACK* st, void* data);
+int OPENSSL_sk_push(_STACK* st, void* data);
+int OPENSSL_sk_unshift(_STACK* st, void* data);
+void* OPENSSL_sk_shift(_STACK* st);
+void* OPENSSL_sk_pop(_STACK* st);
+void OPENSSL_sk_zero(_STACK* st);
+int function(const(void)*, const(void)*) OPENSSL_sk_set_cmp_func(_STACK* sk, ExternC!(int function(const(void)*, const(void)*)) c);
+_STACK* OPENSSL_sk_dup(_STACK* st);
+void OPENSSL_sk_sort(_STACK* st);
+int OPENSSL_sk_is_sorted(const(_STACK)* st);
```

Comment:
Shouldn't the above functions only be in the `DeimosOpenSSLv1.1` block? As far as I could see they don't exist in the earlier versions.

Reply:
It would be more correct, but IMHO having them defined nonetheless might simplify downstream usage  (by removing the need to version if only `stack` is used for example). It seemed to me this approach has upside without visible downside, but I can version it out if you prefer.

Reply:
The question is a bit how backwards compatibility works for OpenSSL. As long as changes are just either permanently removing symbols, or adding new ones with different names, we could try to go this route and support a sliding range of versions. Just as soon as they'd change signatures or things like that we'd be screwed. It also makes it difficult to properly integrate new versions using a three-way diff or by running htod over the latest C headers.

Of course, if these points can be ruled out, having all declarations at once would be the preferable solution.

Reply:
Hum you're right, I moved that code in the `version` block.